iR 594 Audit and add postprocessing effects

MbfloydIR opened the pull request:
Improve PostProcessing API

Added the missing post processing effects

GodRaysEffect needs light source that is not available, so it is left out

The following effects are aded but no difference can be seen when applied

SMAAEffect
SSAOEffect
SSREffect
DepthOfFieldEffect
BloomEffect
SSGIEffect
TRAAEffect
MotionBlurEffect
ShockWaveEffect

MbfloydIR committed:
added color average effect

MbfloydIR committed:
added the outline effect to the post processing

MbfloydIR committed:
added smaa effect for post processing

MbfloydIR committed:
added ChromaticAberrationEffect to post processing

MbfloydIR committed:
added DotScreenEffect post processing effect

MbfloydIR committed:
added TiltShiftEffect post processing effect

MbfloydIR committed:
Added motion blur post processing effect

MbfloydIR committed:
added Glitch Effect to post processing

MbfloydIR committed:
added lineWidth GodRaysEffect and GridEffect to post processing

MbfloydIR committed:
added LUT1DEffect to post processing

MbfloydIR committed:
added LUT3DEffect post processing effect

MbfloydIR committed:
added LUT1DEffect post processing effect

MbfloydIR committed:
added NoiseEffect to post processing

MbfloydIR committed:
added the PixelationEffect to post processing

MbfloydIR committed:
added ScanlineEffect to post processing effects

MbfloydIR committed:
added FXAAEffect and ShockWaveEffect to post processing

MbfloydIR committed:
added TextureEffect to post processing effects

MbfloydIR committed:
added LensDistortionEffect to post processing

MbfloydIR committed:
Merge branch 'dev' into michael/postprocessing

* dev: (42 commits)
  fix(magic-link): Fix guest identity provider removal upon login token validation and retain only the latest generated token (#10078)
  Support GLTF Authoring (#10088)
  fix getNestedChildren not snubing children (#10097)
  Ir 507 Refactor InteractableComponent and expose API (#10064)
  Remove AssetDatabaseType from Asset schema (#10095)
  Added Asset Database Type in schema (#10094)
  Updated argument of install-projects.js script
  Fix distance model deserialization having wrong type check (#10091)
  Add GLTF Scene Support (#10045)
  FileThumbnailJobState now unescapes file names and replaces whitespace characters with hyphens to avoid issues with escaped characters in thumbnail URLs not loading.
  Made changes to input and select components (#10081)
  Fixed a couple bugs in cronjob updating. (#10087)
  tailwind error view and other tailwind fixes (#10076)
  Improve mac os onboarding, disable minio, and update mediasoup (#10071)
  Break Up and Type GLTF Loader (#10044)
  update default scene ground plane to be a nicer colour
  Added check for aws rolearn rollback.
  Added `helm repo update` before `helm upgrade` commands (#10084)
  Added batch invalidation fields to builder helm ugprade command (#10083)
  Modal fixes & UI improvements (#10077)
  ...

# Conflicts:
#	packages/spatial/src/renderer/effects/PostProcessing.ts
#	packages/spatial/src/renderer/functions/configureEffectComposer.ts

HexaField committed:
Merge branch 'dev' into michael/IR-594-postprocessing

DanielBelmes commented:
@MbfloydIR you have some errors on the checks here btw. Double check your commit hooks are running.

MbfloydIR committed:
corrected error with duplicate case option

DanielBelmes commented on packages/editor/src/components/properties/PostProcessingSettingsEditor.tsx:
```diff
@@ -90,7 +97,7 @@ const EffectsOptions: Partial<EffectOptionsType> = {
     fade: { propertyType: PropertyTypes.Number, name: 'Fade', min: -1, max: 1, step: 0.01 },
     resolutionScale: { propertyType: PropertyTypes.Number, name: 'Resolution Scale', min: -10, max: 10, step: 0.01 },
-    kernelSize: { propertyType: PropertyTypes.Number, name: 'Kerne Size', min: 1, max: 5, step: 1 },
+    kernelSize: { propertyType: PropertyTypes.KernelSize, name: 'KernelSize' },
```
The min max and step for this was slightly correct. I actually think there is a KernelSelect we could use for this with values from 0-5. There are others that can use it too.

DanielBelmes commented on packages/spatial/src/renderer/effects/PostProcessing.ts:
```diff
@@ -401,52 +443,52 @@ export type EffectPropsSchema = {
 
 export type EffectPropsSchemaType = (typeof defaultPostProcessingSchema)[keyof typeof defaultPostProcessingSchema]
 
 export const defaultPostProcessingSchema: EffectPropsSchema = {
```
Were you able to check if the default values for the other effects matched the default values from the library/demos?

DanielBelmes commented on packages/spatial/src/renderer/functions/configureEffectComposer.ts:
```diff
@@ -142,27 +161,96 @@ export const configureEffectComposer = (entity: Entity, schema?: typeof defaultP
+      }
+      let texture: Texture | null = null
+      if (texturePath != null) {
+        let textLoad = new TextureLoader()
```
@MichaelEstes Do you have anyway with the resourceManager to load something out of a reactor?

DanielBelmes commented on packages/spatial/src/renderer/functions/configureEffectComposer.ts:
```diff
@@ -142,27 +161,96 @@ export const configureEffectComposer = (entity: Entity, schema?: typeof defaultP
+          const eff = new EffectClass(effectOptions)
+          composer[key] = eff
+          effects.push(eff)
+          AddPass()
```
I believe you're creating multiple duplicates of the EffectPass with `AddPass()`. We have to refactor this logic into PostProcessingComponent so we can use useEffects and the resource manager in order to load these textures.

DanielBelmes reviewed (changes requested):
Some small stuff at first but I believe we do have to refactor configure Effect composer into the PostProcessingComponent earlier than expected for the texture effects.

MbfloydIR committed:
moved the effects composer code into the post processing component so we can use texture loading hooks

HexaField commented on packages/spatial/src/renderer/components/PostProcessingComponent.tsx:
```diff
@@ -63,10 +83,220 @@ export const PostProcessingComponent = defineComponent({
-        rendererEntity,
-        postprocessingComponent.enabled.value ? postprocessingComponent.effects.get(NO_PROXY_STEALTH) : undefined
-      )
+      //configureEffectComposer(
```
we should definitely keep this as a separate function, long useEffects are hard to read

MbfloydIR committed:
committing latest progress for the post processing effects

MbfloydIR committed:
updated the post processing editor to use the lut path

HexaField reviewed (changes requested):
we should add a supreactor for each of the effects, we can then react to them individually, as well as simplify the logic around the texture loading

MbfloydIR committed:
Merge branch 'dev' into michael/IR-594-postprocessing

* dev:
  fix admin projects and build status modals and statuses (#10116)
  Add unique constraint to project name column (#10125)
  9928 optimize asset builds (#9965)
  Physics API Abstraction (#10124)
  rework scene loaded state usage (#10127)
  Camera and Spectation are Event Sourced (#10130)
  Location GLTF State Cleanup (#10129)
  Add locationType menu (#10128)
  IR 465 ECS Materials (#10109)
  Fix scenestate references in playmode tool (#10126)
  Remove old ColliderComponent (#10122)
  GLTF ECS Loader (#10102)
  Systems without any pre/post/sub systems will no longer start out expanded (#10120)
  Updated the avatar radius and camera rotation (#10118)
  Fixed setup helm in dev-deploy (#10092)
  Fix interactable cursor style reactor (#10117)
  IR-1790 making the addComponent search box gain focus automatically when opening the addComponent popup (#10106)
  Update AddEditLocationModal.tsx (#10110)
  Update README.md (#10112)
  MountPoint/Grabbable/Link Components now add Interactables to the authoring layer, optimized useEffects to Systems/Components/Editors (#10105)

MichaelEstes committed:
example

MbfloydIR committed:
moved the effect logic into individual use effects

MbfloydIR committed:
added effect options conditional check

MbfloydIR committed:
removed the associative array map to the effects class since it is not being used in a for loop any longer

HexaField commented on packages/spatial/src/renderer/components/PostProcessingComponent.tsx:
```diff
@@ -78,7 +78,7 @@ export const PostProcessingComponent = defineComponent({
   reactor: () => {
     const entity = useEntityContext()
-    const rendererEntity = useScene(entity)
+    const rendererEntity = useScene(Engine.instance.viewerEntity)
```
why this change? this means any asset panels will not have postprocessing and apply it to the main canvas

MbfloydIR committed:
connected the post processing effects up to the individual useEffect

MbfloydIR committed:
removed the effectMap refs that are no longer needed

MbfloydIR committed:
updated default values for post processing effects
commented out SSGIEffect since it required the composure before it is available

MbfloydIR committed:
created child reactor to wait for the useScene to return a render Entity

MbfloydIR committed:
moving composure declaration so it can be used in an effect constructor for SSREffect

MbfloydIR committed:
alphabetized the effects in the editor

MbfloydIR committed:
added missing SSGIEffect ref

DanielBelmes committed:
fix render pass

DanielBelmes committed:
Merge branch 'dev' into michael/IR-594-postprocessing

HexaField commented:
closing in favour of continued work https://github.com/EtherealEngine/etherealengine/pull/10148
